spec/fake_service/firewall.rb

```diff
@@ -0,0 +1,163 @@
+        params do
+          requires :id, type: Integer
+        end
+        route_param :id do
+          before_validation do
+            @x = $FIREWALLS['firewalls'].find { |x| x['id'].to_s == params[:id].to_s }
+            error!({ error: { code: :not_found } }, 404) if @x.nil?
+          end
+          get do
+            { firewall: @x }
+          end
+
+          group :actions do
+            params do
+              requires :aid, type: Integer
+            end
+            route_param :aid do
+              before_validation do
+                @a = Action.resource_action('firewall', @x['id'], params[:aid])
+                error!({ error: { code: :not_found } }, 404) if @a.nil?
+              end
+              get do
+                { action: @a }
+              end
+            end
+
+            params do
+              optional :status, type: String
+              optional :sort, type: String
+            end
+            get do
+              actions = Action.resource_actions('firewall', @x['id'])
+              unless params[:status].nil?
+                actions['actions'].select! do |x|
+                  x['status'].to_s == params[:status].to_s
+                end
+              end
+              actions
+            end
+
+            params do
+              optional :rules, type: Array
+            end
+            post :set_rules do
+              error!({ error: { code: :invalid_input } }, 400) if params[:rules].nil?
+
+              @x['rules'] = params[:rules]
+
+              a = Action.add(command: 'set_rules', status: 'success',
+                             resources: [{ id: @x['id'], type: 'firewall' }])
+              { action: a }
+            end
+
+            params do
+              optional :apply_to, type: Array
+            end
+            post :apply_to_resources do
+              error!({ error: { code: :invalid_input } }, 400) if params[:apply_to].nil?
+
+              @x['applied_to'].concat(params[:apply_to].map(&:to_h))
+
+              a = Action.add(command: 'apply_to_resources', status: 'success',
+                             resources: [{ id: @x['id'], type: 'firewall' }])
+              { action: a }
+            end
+
+            params do
+              optional :remove_from, type: Array
+            end
+            post :remove_from_resources do
+              error!({ error: { code: :invalid_input } }, 400) if params[:remove_from].nil?
+
+              params[:remove_from].each do |remove_from|
+                @x['applied_to'].delete_if do |applied_to|
+                  (remove_from['type'] == 'server' && applied_to['type'] == 'server' \
+                    && remove_from['server']['id'] == applied_to['server']['id']) \
+                    || \
+                    (remove_from['type'] == 'label_selector' && applied_to['type'] == 'label_selector' \
+                     && remove_from['label_selector']['selector'] == applied_to['label_selector']['selector'])
```

Comment:
quite heavy condition, maybe we can have something more easy to get. Not sure if my variant is much better @RaphaelPour 

```suggestion
                  next unless remove_from['type'] == applied_to['type']
                  
                 case applied_to['type']
                   when 'server'
                     remove_from.dig('server','id') == applied_to.dig('server', 'id')
                   when 'label_selector'
                      remove_from.dig('label_selector','selector') ==  applied_to.dig('label_selector','selector')
                  end
```

Reply:
Yes, I'd prefer Tim's suggestion.